File: andriod/src/CooperHybrid/src/com/codesharp/cooper/Repositories/TasklistRepository.java
package com.codesharp.cooper.Repositories;

import java.util.List;

import android.R.integer;
import android.R.string;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.codesharp.cooper.Constant;
import com.codesharp.cooper.MainActivity;
import com.codesharp.cooper.Tasklist;
import com.codesharp.cooper.TasklistDao;
import com.codesharp.cooper.Tools;
import com.codesharp.cooper.ChangeLogDao.Properties;

public class TasklistRepository {
	
	private MainActivity _activity;
	private TasklistDao _tasklistDao;
	private SharedPreferences _sharedPrefs;
	
	public TasklistRepository(MainActivity activity) {
		this._activity = activity;
		this._tasklistDao = this._activity.getSession().getTasklistDao();
		this._sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this._activity.getApplicationContext());
	}
	
	public List<Tasklist> getAllTasklistByUserAndTemp()
	{
		List<Tasklist> tasklists = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			if(!Tools.isNullOrEmpty(username))
			{
				tasklists  = this._tasklistDao.queryBuilder()
						.where(TasklistDao.Properties.TasklistId.like("temp_%")
								, TasklistDao.Properties.AccountId.eq(username))
							.orderAsc(TasklistDao.Properties.TasklistId)
							.list();
			}
			else
			{
				tasklists  = this._tasklistDao.queryBuilder()
						.where(TasklistDao.Properties.TasklistId.like("temp_%")
								, TasklistDao.Properties.AccountId.eq(""))
							.orderAsc(TasklistDao.Properties.TasklistId)
							.list(); 
			}
		}
		return tasklists;
	}
	
	public List<Tasklist> getAllTasklistByTemp()
	{
		List<Tasklist> tasklists = this._tasklistDao.queryBuilder()
				.where(TasklistDao.Properties.TasklistId.like("temp_%")
						, TasklistDao.Properties.AccountId.eq(""))
				.orderAsc(TasklistDao.Properties.TasklistId)
				.list();
		return tasklists;
	}
	
	public List<Tasklist> getAllTasklist()
	{
		List<Tasklist> tasklists;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			tasklists = this._tasklistDao.queryBuilder()
					.where(TasklistDao.Properties.AccountId.eq(username))
					.orderAsc(TasklistDao.Properties.TasklistId)
					.list();
		}
		else 
		{
			tasklists = this._tasklistDao.queryBuilder()
					.where(TasklistDao.Properties.AccountId.eq(""))
					.orderAsc(TasklistDao.Properties.TasklistId)
					.list();
		}
		return tasklists;
	}
	
	public Tasklist getTasklistById(String tasklistId)
	{
		Tasklist tasklist = null;
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			
			List<Tasklist> tasklists = null;
			if(!Tools.isNullOrEmpty(username))
			{
				tasklists = this._tasklistDao.queryBuilder()
						.where(TasklistDao.Properties.AccountId.eq(username)
								, TasklistDao.Properties.TasklistId.eq(tasklistId))
						.list();
			}
			else
			{
				tasklists = this._tasklistDao.queryBuilder()
						.where(TasklistDao.Properties.AccountId.eq("")
								, TasklistDao.Properties.TasklistId.eq(tasklistId))
						.list();
			}
			if(tasklists.size() > 0)
			{
				tasklist = tasklists.get(0);
			}
		}
		return tasklist;
	}

	public void updateTasklistIdByNewId(String oldId, String newId) {
		
		Tasklist tasklist = this.getTasklistById(oldId);
		if(tasklist != null)
		{
			tasklist.setTasklistId(newId);
		}
		this._tasklistDao.update(tasklist);
	}

	public void deleteAll() {
		
		if(this._sharedPrefs.contains(Constant.USERNAME_KEY))
		{
			String username = this._sharedPrefs.getString(Constant.USERNAME_KEY, "");
			 this._tasklistDao.queryBuilder()
					.where(TasklistDao.Properties.AccountId.eq(username))
					.orderAsc(TasklistDao.Properties.TasklistId)
					.buildDelete()
					.executeDeleteWithoutDetachingEntities();
		}
		else 
		{
			this._tasklistDao.queryBuilder()
					.where(TasklistDao.Properties.AccountId.eq(""))
					.orderAsc(TasklistDao.Properties.TasklistId)
					.buildDelete()
					.executeDeleteWithoutDetachingEntities();
		}
	}

	public void addTasklist(Tasklist tasklist) {
		this._tasklistDao.insert(tasklist);
	}

	public void addTasklists(List<Tasklist> tasklists) {
		this._tasklistDao.insertInTx(tasklists);
	}

	public void updateEditable(int editable, String tasklistId) {
		Tasklist tasklist = this.getTasklistById(tasklistId);
		if(tasklist != null)
		{
			tasklist.setEditable(editable == 1 ? true : false);
		}
		this._tasklistDao.update(tasklist);
	}

	//批量更新任务列表
	public void updateTasklists(List<Tasklist> tasklists) {
		
		this._tasklistDao.updateInTx(tasklists);
	}
}

File: windows phone/src/Cooper.Repositories/TasklistRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using Cooper.Core;
using Cooper.Core.Models;
using System.IO.IsolatedStorage;

namespace Cooper.Repositories
{
    public class TasklistRepository
    {
        private CooperContext _context;

        public TasklistRepository()
        {
            this._context = new CooperContext();
        }

        public List<Tasklist> GetAllTasklistByUserAndTemp()
        {
            List<Tasklist> tasklists = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                tasklists = this._context.Tasklists.Where(o => o.TasklistId.IndexOf("temp_") >= 0
                    && o.AccountId.Equals(username))
                    .OrderBy(o => o.TasklistId)
                    .ToList();
            }
            else
            {
                tasklists = this._context.Tasklists.Where(o => o.TasklistId.IndexOf("temp_") >= 0
                    && string.IsNullOrEmpty(o.AccountId))
                    .OrderBy(o => o.TasklistId)
                    .ToList();
            }
            return tasklists;
        }

        public List<Tasklist> GetAllTasklistByTemp()
        {
            List<Tasklist> tasklists = this._context.Tasklists
                .Where(o => o.TasklistId.IndexOf("temp_") >= 0
                    && o.AccountId.Equals(""))
                    .ToList();
            return tasklists;
        }

        public List<Tasklist> GetAllTasklist()
        {
            List<Tasklist> tasklists = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                tasklists = this._context.Tasklists.Where(o => o.AccountId.Equals(username))
                    .OrderBy(o => o.TasklistId)
                    .ToList();
            }
            else
            {
                tasklists = this._context.Tasklists.Where(o => o.AccountId.Equals(""))
                    .OrderBy(o => o.TasklistId)
                    .ToList();
            }
            return tasklists;
        }

        public Tasklist GetTasklistById(string tasklistId)
        {
            Tasklist tasklist = null;
            List<Tasklist> tasklists = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                tasklists = this._context.Tasklists.Where(o => o.AccountId.Equals(username)
                    && o.TasklistId.Equals(tasklistId))
                    .ToList();
            }
            else
            {
                tasklists = this._context.Tasklists.Where(o => o.AccountId.Equals("")
                   && o.TasklistId.Equals(tasklistId))
                   .ToList();
            }
            if (tasklists != null && tasklists.Count > 0)
            {
                tasklist = tasklists[0];
            }
            return tasklist;
        }

        public void UpdateTasklistByNewId(string oldId, string newId)
        {
            Tasklist tasklist = this.GetTasklistById(oldId);
            if (tasklist != null)
            {
                tasklist.TasklistId = newId;
            }
            this._context.SubmitChanges();
        }

        public void DeleteAll()
        {
            List<Tasklist> tasklists = null;
            if (IsolatedStorageSettings.ApplicationSettings.Contains(Constant.USERNAME_KEY))
            {
                string username = IsolatedStorageSettings.ApplicationSettings[Constant.USERNAME_KEY] as string;
                tasklists = this._context.Tasklists.Where(o => o.AccountId.Equals(username)).ToList();
            }
            else
            {
                tasklists = this._context.Tasklists.Where(o => o.AccountId.Equals("")).ToList();
            }
            if (tasklists != null && tasklists.Count > 0)
            {
                this._context.Tasklists.DeleteAllOnSubmit<Tasklist>(tasklists);
                this._context.SubmitChanges();
            }   
        }

        public void AddTasklist(Tasklist tasklist)
        {
            this._context.Tasklists.InsertOnSubmit(tasklist);
            this._context.SubmitChanges();
        }

        public void AddTasklists(List<Tasklist> tasklists)
        {
            this._context.Tasklists.InsertAllOnSubmit<Tasklist>(tasklists);
            this._context.SubmitChanges();
        }

        public void UpdateEditable(int editable, string tasklistId)
        {
            Tasklist tasklist = this.GetTasklistById(tasklistId);
            if (tasklist != null)
            {
                tasklist.Editable = editable == 1;
            }
            this._context.SubmitChanges();
        }

        public void UpdateTasklists(List<Tasklist> tasklists)
        {
            this._context.SubmitChanges();
        }

        public void AdjustWithNewId(string oldId, string newId)
        {
            Tasklist tasklist = null;
            List<Tasklist> tasklists = this._context.Tasklists.Where(o => o.TasklistId.Equals(oldId)).ToList();
            if (tasklists.Count > 0)
            {
                tasklist = tasklists[0];
                tasklist.TasklistId = newId;
            }
            this._context.SubmitChanges();
        }
    }
}
